io/src/lib.rs

```diff
@@ -230,6 +230,57 @@ impl<T: AsRef<[u8]>> BufRead for Cursor<T> {
     }
 }
 
+/// A [`Read`]er which keeps an internal buffer to avoid hitting the underlying stream directly for
+/// every read, implementing [`BufRead`].
+pub struct BufReader<'a, R: Read> {
+    inner: &'a mut R,
+    buf: [u8; 4096],
+    pos: usize,
+    limit: usize,
+}
+
+impl<'a, R: Read> BufReader<'a, R> {
+    /// Creates a [`BufReader`] which will read from the given `inner`.
+    pub fn new(inner: &'a mut R) -> Self {
+        BufReader {
+            inner,
+            buf: [0; 4096],
```

Comment:
There's some needless zeroing, ideally we would use `ArrayVec<u8>` but sadly we still can't because our read API doesn't support reading into uninitialized buffers (just like std API). We could fix it just like `std` did in nightly but not sure if it's worth it.